lib/pagy/backend.rb

```diff
@@ -16,10 +16,17 @@ def pagy(collection, vars={})
 
     # Sub-method called only by #pagy: here for easy customization of variables by overriding
     def pagy_get_vars(collection, vars)
-      count = collection.count(:all)                # work with AR collections: other ORMs may need to change this
-      count = count.count if count.is_a?(Hash)      # fix for the AR grouping count inconsistency (Hash instead of Integer)
+      count =
+        if vars.key?(:count)
+          vars[:count]
+        elsif collection.group_values.any?
+          collection.count(:all).count  # for the AR grouping count inconsistency (Hash instead of Integer)
```

Comment:
for big performance this code may replaced follow
```ruby
ActiveRecord::Base.connection.select_value(<<~SQL)
  SELECT count(*) FROM (#{collection.except(:select).select('count(*)').to_sql}) t
SQL
```
but requires for test follow dependencies

```
 spec.add_development_dependency 'combustion'
 spec.add_development_dependency 'sqlite3'
 spec.add_development_dependency 'activerecord'
```

Reply:
```ruby
require 'benchmark/ips'
scope = Order.where("created_at > '2018-05-05'").group(:user_id)
Benchmark.ips do |x|
  x.report("sub sql") { 
	ActiveRecord::Base.connection.select_value(<<~SQL)
	  SELECT count(*) FROM (#{scope.except(:select).select('count(*)').to_sql}) t
	SQL
  }
  x.report("hash count") { scope.count(:all).count }
  x.report("hash size") { scope.count(:all).size }
  x.compare!
end
```
```
Calculating -------------------------------------
             sub sql      6.291  (± 0.0%) i/s -     32.000  in   5.114788s
          hash count      2.990  (± 0.0%) i/s -     15.000  in   5.028347s
           hash size      2.865  (± 0.0%) i/s -     15.000  in   5.300102s

Comparison:
             sub sql:        6.3 i/s
          hash count:        3.0 i/s - 2.10x  slower
           hash size:        2.9 i/s - 2.20x  slower
```

```ruby
grouped_orders = Order.where("created_at > '2018-05-05'").group(:user_id).count(:all)
Benchmark.ips do |x|
  x.report("hash count") { grouped_orders.count }
  x.report("hash size") { grouped_orders.size }
  x.compare!
end
```

```
Warming up --------------------------------------
          hash count    41.000  i/100ms
           hash size   356.849k i/100ms
Calculating -------------------------------------
          hash count    404.438  (± 9.9%) i/s -      2.009k in   5.021962s
           hash size     11.470M (± 4.4%) i/s -     57.453M in   5.019445s

Comparison:
           hash size: 11469606.0 i/s
          hash count:      404.4 i/s - 28359.35x  slower
```

sub sql 2x times faster, in addition, it uses less memory because it does not create a large hash

Reply:
I tried it quickly but it couldn't make it work. Is it only for group or for everything? Maybe you could write a small section in the [Dealing with a slow count...](https://ddnexus.github.io/pagy/how-to#dealing-with-a-slow-collection-count), explaining the details...

Reply:
where and how did you try? could cite some code

sub sql only for queries with grouping, it should work in postgres, mysql, sqlite
ok I will try to write an instruction

also note that the hash.count is very slow, benchmark above

Reply:
You are totally right! I completely forgot that `count` is meant to iterate over the results of a block and without any block it still counts by iterating. Very expensive! I will change that. Thank you!